### DISCUSSION TITLE
Poor Performance

### DISCUSSION THREAD

**Discussion by guibirow** [saleor/saleor#8754] [General]
Hi All,

My company implemented saleor 2.11 a while ago for our catalogue and order management system(not the store front) and the performance metrics are very bad that makes me wonder if we did something wrong when deploying it or it should be expected.

We currently have 32 instances of saleor deployed with kubernetes on multiple nodes on google cloud.
Each intance has allocated 1GB RAM and 1vCPU (utilised 800Mb ram and 0.2 CPU)
The database is a Cloud SQL Postgress Instance with 4vCPU and 16GB Ram

All the instances have an NGINX in front of it to do some basic request transformation(like headers removal) and is exposed to the internet from a Google Load Balancer.

All these instances, can handle together under normal operation, around 20-30 requests per second with the following metrics:
SQL CPU Utilization: 70%
SQL Memory: 6Gb
Response Time: 800ms (median) | 3000ms (95th)

Adding more instances increases the throughput slighly but not much, about 1-4 requests per additional instance.
At peek, the maximun number of requests it can handled before starting throwing exceptions or timing out is about 50req/s
Based on all these metrics, we can get a rough estimation of 1 request per instance.

Are these values expected? 
If not, could it be some misconfiguration we have done(very likely, but I can't find what)?
What should be the expected reponse times and requests per second based on any reference setup you have in production today.

The person who initially implemented it is not in the company anymore, so we can't for sure say if this has been degrading with time or was this bad since the beginning.

**Comment by patrys** [comment]
@guibirow, was the version you use customized in any way? That performance profile does not sound right. A lot depends on the exact GraphQL queries you're executing and whether you're requesting only the necessary information.

While I can't help you directly with your codebase, I can recommend looking into upgrading to Saleor 3.0 when time and resources allow. We made a lot of improvements to Saleor since 2.11, and 3.0 is better in every regard, including performance.

But then again, your existing service is much slower than what I'd expect to see, and the high SQL CPU use seems to indicate that the problem is with the amount of data you fetch with every query.

**Reply by guibirow** [reply]
It was customized to add new features, but not big changes to be honest, mainly adding new fields and additional tables for management of new features in the dashboard.

Main load comes from a mobile app used to browse the catalogue. We don't use the saleor website.

AFAIK, the mobile app request only the necessary information, but the queries in the SQL seems to be loading a lot of data.
I suspect Saleor loads everything from a table even for small requests where we need only 1 field. I will setup a new environment to investigate this point.

I had a chat with the developers and they mentioned each instance is limited to 4 parallel requests, but they couldn't explain why properly.
They did a load test and could only handle 4 request per instance at peak.
They suspect it is something to do with Django and the framework used to handle the webserver.
Does it make any sense?

I can see in the docker file there is 4 workers setup, but is the same as the one in the original repo.
``` CMD ["gunicorn", "--bind", ":8000", "--workers", "4", "--worker-class", "uvicorn.workers.UvicornWorker", "saleor.asgi:application"]```

I wil ask them to investigate how much effort it will be to upgrade to 3.0. Thanks for the advice!

**Reply by korycins** [reply]
Hey @guibirow , Not sure if this is something related to your issue, but still I would double check this: In 2.11.9, we released a performance fix for loading a full DB model to app memory, when graphql query didn't have an input parameter: `first`, `last`.
https://github.com/saleor/saleor/releases/tag/2.11.9
